File: chroma-league/src/main/java/com/bonepl/chromaleague/hud/animations/RespawnAnimation.java
package com.bonepl.chromaleague.hud.animations;

import com.bonepl.chromaleague.hud.colors.BreathingColor;
import com.bonepl.chromaleague.hud.colors.CLColor;
import com.bonepl.chromaleague.hud.colors.TransitionColor;
import com.bonepl.chromaleague.hud.parts.Background;
import com.bonepl.chromaleague.hud.parts.health.HealthBar;
import com.bonepl.chromaleague.hud.parts.resource.ResourceBars;
import com.bonepl.chromaleague.hud.parts.resource.ShyvanaDragonFuryBar;
import com.bonepl.chromaleague.hud.parts.resource.YasuoWindBar;
import com.bonepl.chromaleague.state.GameStateHelper;
import com.bonepl.chromaleague.state.RunningState;
import com.bonepl.razersdk.animation.AnimatedFrame;
import com.bonepl.razersdk.animation.Color;
import com.bonepl.razersdk.animation.LayeredFrame;
import com.bonepl.razersdk.animation.SimpleFrame;
import com.bonepl.razersdk.sdk.RzKey;
import com.bonepl.razersdk.sdk.RzKeySelector;

import java.util.Arrays;
import java.util.List;
import java.util.Set;

import static com.bonepl.razersdk.sdk.RzKey.*;

public class RespawnAnimation extends AnimatedFrame {
    private static final Set<RzKey> GLOW_AREA = BaronBuffBackgroundAnimation.buildBaronArea();
    private static final List<RzKey> FOURTH_ROW = new RzKeySelector().withRowOf(RZKEY_Q).withColumnBetween(RZKEY_Q, RZKEY_U).asList();
    private static final List<RzKey> THIRD_ROW = new RzKeySelector().withRowOf(RZKEY_A).withColumnBetween(RZKEY_A, RZKEY_H).asList();
    private static final List<RzKey> SECOND_ROW = new RzKeySelector().withRowOf(RZKEY_Z).withColumnBetween(RZKEY_Z, RZKEY_B).asList();
    private static final List<RzKey> FIRST_ROW = new RzKeySelector().withRowOf(RZKEY_LALT).withColumnBetween(RZKEY_LALT, RZKEY_SPACE).asList();
    private static final int STEPS = 20;

    public RespawnAnimation() {
        BreathingColor glowAreaColor = new BreathingColor(Color.YELLOW, STEPS, true);
        BreathingColor buttonsColor = new BreathingColor(Color.YELLOW, STEPS, true);
        int delayBetweenRows = 5;
        final List<AnimatedFrame> animatedFrames = Arrays.asList(
                createYellowAnimatedFrame(0, FIRST_ROW),
                createYellowAnimatedFrame(delayBetweenRows << 1, SECOND_ROW),
                createYellowAnimatedFrame(delayBetweenRows * 3, THIRD_ROW),
                createYellowAnimatedFrame(delayBetweenRows << 2, FOURTH_ROW),
                createButtonsGlowAnimatedFrame(delayBetweenRows * 5, ResourceBars.getResourceBarKeys(), delayBetweenRows * 7),
                createButtonsGlowAnimatedFrame(delayBetweenRows * 6, HealthBar.getHealthBarKeys(), delayBetweenRows * 7),
                createButtonsTransitionAnimatedFrame(delayBetweenRows * 7, ResourceBars.getResourceBarKeys(), getResourceColor()),
                createButtonsTransitionAnimatedFrame(delayBetweenRows * 7, HealthBar.getHealthBarKeys(), Color.GREEN)
        );
        for (int i = 0; i < delayBetweenRows * 7 + (STEPS << 1); i++) {
            final LayeredFrame layeredFrame = new LayeredFrame();
            animatedFrames.stream().filter(AnimatedFrame::hasFrame).forEach(layeredFrame::addFrame);
            addAnimationFrame(layeredFrame);
        }
    }

    private static Color getResourceColor() {
        Color toResourceColor;
        if (GameStateHelper.getResourcePercentage() == 0) {
            toResourceColor = Background.BACKGROUND_COLOR;
        } else {
            toResourceColor = getPlayerResourceToTransitionColor();
        }
        return toResourceColor;
    }

    private static AnimatedFrame createYellowAnimatedFrame(int delay, List<RzKey> keys) {
        final AnimatedFrame animatedFrame = new AnimatedFrame();
        animatedFrame.addAnimationFrame(delay, new SimpleFrame());
        BreathingColor yellowBreathingColor = new BreathingColor(Color.YELLOW, STEPS, true);
        for (int i = 0; i < STEPS << 1; i++) {
            final LayeredFrame layeredFrame = new LayeredFrame();
            layeredFrame.addFrame(new SimpleFrame(keys, yellowBreathingColor.getNextColor()));
            if (i < STEPS) {
                layeredFrame.addFrame(new SimpleFrame(HealthBar.getHealthBarKeys(), Background.BACKGROUND_COLOR));
                layeredFrame.addFrame(new SimpleFrame(ResourceBars.getResourceBarKeys(), Background.BACKGROUND_COLOR));
            }
            animatedFrame.addAnimationFrame(layeredFrame);
        }
        return animatedFrame;
    }

    private static AnimatedFrame createButtonsGlowAnimatedFrame(int delay, List<RzKey> keys, int waitTill) {
        final AnimatedFrame animatedFrame = new AnimatedFrame();
        animatedFrame.addAnimationFrame(delay, new SimpleFrame());
        BreathingColor yellowBreathingColor = new BreathingColor(Color.YELLOW, STEPS, true);
        for (int i = 0; i < STEPS; i++) {
            animatedFrame.addAnimationFrame(new SimpleFrame(keys, yellowBreathingColor.getNextColor()));
        }
        for (int i = delay + (STEPS << 1); i < waitTill - delay; i++) {
            animatedFrame.addAnimationFrame(new SimpleFrame(keys, Color.YELLOW));
        }
        return animatedFrame;
    }

    private static AnimatedFrame createButtonsTransitionAnimatedFrame(int delay, List<RzKey> keys, Color toColor) {
        final AnimatedFrame animatedFrame = new AnimatedFrame();
        animatedFrame.addAnimationFrame(delay, new SimpleFrame());
        TransitionColor buttonTransitionColor = new TransitionColor(Color.YELLOW, toColor, STEPS);
        for (int i = 0; i < STEPS; i++) {
            animatedFrame.addAnimationFrame(new SimpleFrame(keys, buttonTransitionColor.getNextColor()));
        }
        return animatedFrame;
    }

    private static Color getPlayerResourceToTransitionColor() {
        String activePlayerChampionName = RunningState.getGameState().getPlayerList().getActivePlayer().getChampionName();
        if (ResourceBars.getEnergyBarChampions().contains(activePlayerChampionName)) {
            return Color.YELLOW;
        }
        if ("Shyvana".equals(activePlayerChampionName)) {
            return ShyvanaDragonFuryBar.DRAGON_FURY_COLOR;
        }
        if ("Yasuo".equals(activePlayerChampionName)) {
            return YasuoWindBar.WIND_SHIELD_COLOR;
        }
        return CLColor.MANA;
    }
}

File: src/hud/animations/RespawnAnimation.ts
import {AnimatedFrame} from '../../razer-sdk/animation/AnimatedFrame.js';
import {LayeredFrame} from '../../razer-sdk/animation/LayeredFrame.js';
import {SimpleFrame} from '../../razer-sdk/animation/SimpleFrame.js';
import {BreathingColor} from '../../razer-sdk/color/BreathingColor.js';
import {StaticColor} from '../../razer-sdk/color/StaticColor.js';
import {TransitionColor} from '../../razer-sdk/color/TransitionColor.js';
import {RzKey} from '../../razer-sdk/sdk/RzKey.js';
import {RzKeySelector} from '../../razer-sdk/sdk/RzKeySelector.js';
import {CLColor} from '../colors/CLColor.js';
import {BackgroundBreathingColor} from '../colors/BackgroundBreathingColor.js';
import {getCurrentBackgroundColor} from '../parts/Background.js';
import {getHealthBarKeys} from '../parts/health/HealthBar.js';
import {getEnergyBarChampions, getResourceBarKeys} from '../parts/resource/ResourceBars.js';
import {DRAGON_FURY_COLOR} from '../parts/resource/ShyvanaDragonFuryBar.js';
import {WIND_SHIELD_COLOR} from '../parts/resource/YasuoWindBar.js';
import * as GameStateHelper from '../../state/GameStateHelper.js';
import {RunningState} from '../../state/RunningState.js';

const FOURTH_ROW: RzKey[] = new RzKeySelector().withRowOf(RzKey.RZKEY_Q).withColumnBetween(RzKey.RZKEY_Q, RzKey.RZKEY_U).asList();
const THIRD_ROW: RzKey[] = new RzKeySelector().withRowOf(RzKey.RZKEY_A).withColumnBetween(RzKey.RZKEY_A, RzKey.RZKEY_H).asList();
const SECOND_ROW: RzKey[] = new RzKeySelector().withRowOf(RzKey.RZKEY_Z).withColumnBetween(RzKey.RZKEY_Z, RzKey.RZKEY_B).asList();
const FIRST_ROW: RzKey[] = new RzKeySelector().withRowOf(RzKey.RZKEY_LALT).withColumnBetween(RzKey.RZKEY_LALT, RzKey.RZKEY_SPACE).asList();
const STEPS = 20;

function getResourceColor(): StaticColor {
    if (GameStateHelper.getResourcePercentage() === 0) {
        return new StaticColor(5, 5, 5); // DEFAULT_BACKGROUND_COLOR
    }
    return getPlayerResourceToTransitionColor();
}

function getPlayerResourceToTransitionColor(): StaticColor {
    const activePlayerChampionName = RunningState.getGameState().playerList!
        .getActivePlayer(RunningState.getGameState().playerRiotId!).championName;
    if (getEnergyBarChampions().has(activePlayerChampionName)) {
        return StaticColor.YELLOW;
    }
    if (activePlayerChampionName === 'Shyvana') {
        return DRAGON_FURY_COLOR;
    }
    if (activePlayerChampionName === 'Yasuo') {
        return WIND_SHIELD_COLOR;
    }
    return CLColor.MANA;
}

function createYellowAnimatedFrame(delay: number, keys: RzKey[]): AnimatedFrame {
    const animatedFrame = new AnimatedFrame();
    animatedFrame.addAnimationFrame(delay, new SimpleFrame());
    const yellowBreathingColor: BreathingColor = new BackgroundBreathingColor(StaticColor.YELLOW, STEPS, true);
    for (let i = 0; i < STEPS << 1; i++) {
        const layeredFrame = new LayeredFrame();
        layeredFrame.addFrame(new SimpleFrame([...keys], yellowBreathingColor.getColor()));
        if (i < STEPS) {
            layeredFrame.addFrame(new SimpleFrame([...getHealthBarKeys()], getCurrentBackgroundColor()));
            layeredFrame.addFrame(new SimpleFrame([...getResourceBarKeys()], getCurrentBackgroundColor()));
        }
        animatedFrame.addAnimationFrame(layeredFrame);
    }
    return animatedFrame;
}

function createButtonsGlowAnimatedFrame(delay: number, keys: readonly RzKey[], waitTill: number): AnimatedFrame {
    const keyArr = [...keys];
    const animatedFrame = new AnimatedFrame();
    animatedFrame.addAnimationFrame(delay, new SimpleFrame());
    const yellowBreathingColor: BreathingColor = new BackgroundBreathingColor(StaticColor.YELLOW, STEPS, true);
    for (let i = 0; i < STEPS; i++) {
        animatedFrame.addAnimationFrame(new SimpleFrame(keyArr, yellowBreathingColor.getColor()));
    }
    for (let i = delay + (STEPS << 1); i < waitTill - delay; i++) {
        animatedFrame.addAnimationFrame(new SimpleFrame(keyArr, StaticColor.YELLOW));
    }
    return animatedFrame;
}

function createButtonsTransitionAnimatedFrame(delay: number, keys: readonly RzKey[], toColor: StaticColor): AnimatedFrame {
    const keyArr = [...keys];
    const animatedFrame = new AnimatedFrame();
    animatedFrame.addAnimationFrame(delay, new SimpleFrame());
    const buttonTransitionColor = new TransitionColor(StaticColor.YELLOW, toColor, STEPS);
    for (let i = 0; i < STEPS; i++) {
        animatedFrame.addAnimationFrame(new SimpleFrame(keyArr, buttonTransitionColor.getColor()));
    }
    return animatedFrame;
}

export class RespawnAnimation extends AnimatedFrame {
    constructor() {
        super();
        const delayBetweenRows = 5;
        const animatedFrames: AnimatedFrame[] = [
            createYellowAnimatedFrame(0, FIRST_ROW),
            createYellowAnimatedFrame(delayBetweenRows << 1, SECOND_ROW),
            createYellowAnimatedFrame(delayBetweenRows * 3, THIRD_ROW),
            createYellowAnimatedFrame(delayBetweenRows << 2, FOURTH_ROW),
            createButtonsGlowAnimatedFrame(delayBetweenRows * 5, getResourceBarKeys(), delayBetweenRows * 7),
            createButtonsGlowAnimatedFrame(delayBetweenRows * 6, getHealthBarKeys(), delayBetweenRows * 7),
            createButtonsTransitionAnimatedFrame(delayBetweenRows * 7, getResourceBarKeys(), getResourceColor()),
            createButtonsTransitionAnimatedFrame(delayBetweenRows * 7, getHealthBarKeys(), StaticColor.GREEN),
        ];
        for (let i = 0; i < delayBetweenRows * 7 + (STEPS << 1); i++) {
            const layeredFrame = new LayeredFrame();
            for (const af of animatedFrames) {
                if (af.hasFrame()) {
                    layeredFrame.addFrame(af);
                }
            }
            this.addAnimationFrame(layeredFrame);
        }
    }
}
